### PR TITLE
Assignment7+8 linq

### PR DESCRIPTION
**Pull request by JosephPotapenko**
Worked on this with @steeley21.

### CODE REVIEW

File: Assignment/Assignment/SampleData.cs [hunk 1]
```diff
@@ -1,29 +1,93 @@
...
+    // 1. 
+    public IEnumerable<string> CsvRows
+    {
+        get
+        {
+            string baseDir = AppContext.BaseDirectory;
+            string csvPath = Path.Combine(baseDir, "People.csv");
+            if (!File.Exists(csvPath))
+            {
+                string probe = Path.Combine(baseDir, "..", "..", "..", "Assignment", "Assignment", "People.csv");
```
Comment (thigiang16) [inline Assignment/Assignment/SampleData.cs]: I think if “Copy if newer” is set correctly on the file, it should always appear in the output directory, so you wouldn’t need to search for it manually. Did the file not copy over for you, or was there another issue you were trying to handle?

File: Assignment/Assignment/SampleData.cs [hunk 2]
```diff
@@ -1,29 +1,93 @@
...
 
-    // 4.
-    public IEnumerable<IPerson> People => throw new NotImplementedException();
+    // 6. 
+    public string GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)
+    {
+        HashSet<string> seen = new(StringComparer.OrdinalIgnoreCase);
+        string[] ordered = people.Select(p => p.Address.State)
+            .Where(s => !string.IsNullOrWhiteSpace(s))
+            .Where(s => seen.Add(s)).ToArray();
```
Comment (thigiang16) [inline Assignment/Assignment/SampleData.cs]: I think the assignment expects this to be done with LINQ, so you don’t need the manual HashSet. I used Distinct, OrderBy, and Aggregate for mine. Also, you don’t need ToArray() here, the sequence can stay lazy as an IEnumerable.

File: Assignment.Tests/NodeTests.cs [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+        Node<int> n1 = new Node<int>(1);
+        n1.Append(2);
+        n1.Append(3);
+        
+        // Act
+        List<int> result = n1.ChildItems(0).ToList();
+
+        // Assert
+        Assert.IsEmpty(result);
+    }
```
Comment (thigiang16) [inline Assignment.Tests/NodeTests.cs:266]: I suggest adding tests to cover GetEnumerator by directly iterating through the node using foreach or ToList(), and also adding a test for ChildItems on an empty node.

File: Assignment.Tests/SampleDataTests.cs [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+[TestClass]
+public class SampleDataTests
+{
+    private static readonly string[] expected = ["OR", "WA"];
+
+    [TestMethod]
+    public void GetUniqueSortedListOfStates_HardcodedDuplicateStates_ReturnsUniqueSorted()
+    {
+        string[] csvRows =
```
Comment (thigiang16) [inline Assignment.Tests/SampleDataTests.cs:16]: Most of your arrays are defined using [...]. I think new[] { ... } should be used instead for correct C# syntax

**Review by thigiang16 — COMMENTED** [review]
Implement the ISampleData.CsvRows property, loading the data from the People.csv file and returning each line as a single string. ✔

Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ❌
Using LINQ, skip the first row in the People.csv. ✔
Be sure to appropriately handle resource (IDisposable) items correctly if applicable (and it may not be depending on how you implement it). ✔
Implement IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows() to return a sorted, unique list of states. ✔

Use ISampleData.CsvRows for your data source. ✔
Don't forget the list should be unique. ✔
Sort the list alphabetically. ✔
Include a test that leverages a hardcoded list of addresses. ✔
Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔
Implement ISampleData.GetAggregateSortedListOfStatesUsingCsvRows() to return a string that contains a unique, comma separated list of states. ✔

Use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows() for your data source. ✔
Consider "selecting" only the states and calling ToArray() to retrieve an array of all the state names. ✔
Given the array, consider using string.Join to combine the list into a single string.✔
Implement the ISampleData.People property to return all the items in People.csv as Person objects ✔

Use ISampleData.CsvRows as the source of the data. ✔
Sort the list by State, City, and Zip. ✔
Be sure that Person.Address is also populated. ✔
Adding null validation to all the Person and Address properties is optional.
Consider using ISampleData.CsvRows in your test to verify your results. ✔
Implement ISampleDate.FilterByEmailAddress(Predicate<string> filter) to return a list of names where the email address matches the filter. ✔

Use ISampleData.People for your data source. ✔
Implement ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people) to return a string that contains a unique, comma-separated list of states. ✔

Use the people parameter from ISampleData.People property for your data source. ✔
At a minimum, use the System.Linq.Enumerable.Aggregate` LINQ method to create your result. ✔
Don't forget the list should be unique. ✔
It is recommended that, at a minimum, you use ISampleData.GetUniqueSortedListOfStatesGivenCsvRows to validate your result.
Given the implementation of Node in Assignment5

Implement IEnumerable<T> to return all the items in the "circle" of items. ✔
Add an IEnumberable<T> ChildItems(int maximum) method to Node that returns the remaining items with a maximum number of items returned less than maximum.
Extra Credit
Implement the homework using async/await and multi-threading by defining a new SampleDataAsync class that implements IAsyncSampleData). Refactor your SampleData and SampleDataAsync classes with minimal duplication. Be sure to refactor your tests to re-use a significant amount of the test code for both implementations. ❌
Fundamentals
Place all shared project properties into a Directory.Build.Props file.
Place all shared project items into a Directory.Build.targets file.
Ensure nullable reference types is enabled ✔
Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
Set LangVersion and the TargetFramework to the latest released versions available (preview versions optional) ✔
and enabled .NET analyzers for both projects ✔
For this assignment, consider using Assert.AreEqual<T>() (the generic version) ✔
All of the above should be unit tested ❌
Choose simplicity over complexity ✔

File: Assignment/Assignment/SampleData.cs [hunk 5]
```diff
@@ -1,29 +1,93 @@
...
+            string baseDir = AppContext.BaseDirectory;
+            string csvPath = Path.Combine(baseDir, "People.csv");
+            if (!File.Exists(csvPath))
+            {
+                string probe = Path.Combine(baseDir, "..", "..", "..", "Assignment", "Assignment", "People.csv");
+                string candidate = Path.GetFullPath(probe);
+                if (File.Exists(candidate)) csvPath = candidate;
+                else throw new FileNotFoundException($"People.csv not found at '{csvPath}'.");
+            }
+            return File.ReadLines(csvPath).Skip(1).ToArray();
```
[the remark this answers is not in the document]
Comment (JosephPotapenko) [reply Assignment/Assignment/SampleData.cs:21]: While yes, you aren't implementing lazy enumeration- but doing .ToArray() makes sure the data is fixed and reusable without re-reading the file.

File: Assignment/Assignment/Node.cs [hunk 6]
```diff
@@ -0,0 +1,237 @@
...
+    {
+        ArgumentNullException.ThrowIfNull(array);
+        ArgumentOutOfRangeException.ThrowIfNegative(arrayIndex, nameof(arrayIndex));
+        if (array.Length - arrayIndex < Count) throw new ArgumentException("The destination array has insufficient space.");
+        if (_IsEmpty) return;
+        array[arrayIndex++] = _Value;
+
+        if (ReferenceEquals(this.Next, this)) return;
+        int appendedCount = Count - 1;
+        int writeIndex = arrayIndex + appendedCount - 1;
```
[the remark this answers is not in the document]
Comment (JosephPotapenko) [reply Assignment/Assignment/Node.cs:147]: The code already handles the arrayIndex increment on line 143. Then line 147 uses that updated value to figure out exactly where the remaining items go

File: Assignment/Assignment/SampleData.cs (hunk 1 above)
Comment (JosephPotapenko) [reply Assignment/Assignment/SampleData.cs]: You're totally right on that. I coded the first part of this assignment in the files solution and end up running the program outside the .sln which caused the issue where it didn't copy over. Totally my fault. I mean, I could leave it to handle a potential edge case like what I ran into, but I think you're right and I should probably get rid of it.

Comment (JosephPotapenko) [reply Assignment/Assignment/SampleData.cs]: Just got rid of it and everything still works like a charm. Thank-you 👍

File: Assignment/Assignment/SampleData.cs (hunk 2 above)
Comment (JosephPotapenko) [reply Assignment/Assignment/SampleData.cs]: You're totally right. Thanks for the advice! It made implementation a lot easier. 👍

File: Assignment.Tests/NodeTests.cs (hunk 3 above)
Comment (steeley21) [reply Assignment.Tests/NodeTests.cs:266]: Good call! These tests have been added

File: Assignment.Tests/SampleDataTests.cs (hunk 4 above)
Comment (steeley21) [reply Assignment.Tests/SampleDataTests.cs:16]: Thanks for your comment! The syntax here is using collection expressions, which are valid for initializing arrays and other collections. The way you suggest would also be correct, but since this syntax is also valid we’ll keep it as is.

Here is a Microsoft article about collection expressions and how to use them if you were curious: https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/collection-expressions

Comment (thigiang16) [reply Assignment.Tests/SampleDataTests.cs:16]: Thanks for the clarification and the link! 😄

**Review by Joshua-Lester3 — COMMENTED** [review]
## Instructions

**Throughout, consider using the `System.Linq.Enumerable` methods `Zip`, `Count`, `Sort` and `Contains` methods for testing collections.**. (Preferably avoid using `Microsoft.VisualStudio.TestTools.UnitTesting.CollectionAssert` although that might be easier, to get a firmer grasp on additional LINQ API.)

1. Implement the `ISampleData.CsvRows` property, loading the data from the `People.csv` file and returning each line as a single string. ✔

   - Change the "Copy to" property on People.csv to "Copy if newer" so that the file is deployed along with your test project. ✔
   - Using LINQ, skip the first row in the `People.csv`. ✔
   - Be sure to appropriately handle resource (`IDisposable`) items correctly if applicable (and it may not be depending on how you implement it). ✔

2. Implement `IEnumerable<string> GetUniqueSortedListOfStatesGivenCsvRows()` to return a **sorted**, **unique** list of states. ✔

   - Use `ISampleData.CsvRows` for your data source. ✔
   - Don't forget the list should be unique. ✔
   - Sort the list alphabetically. ✔
   - Include a test that leverages a hardcoded list of addresses. ✔
   - Include a test that uses LINQ to verify the data is sorted correctly (do not use a hardcoded list). ✔

3. Implement `ISampleData.GetAggregateSortedListOfStatesUsingCsvRows()` to return a `string` that contains a **unique**, comma separated list of states. ✔

   - Use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows()` for your data source. ✔
   - Consider "selecting" only the states and calling `ToArray()` to retrieve an array of all the state names. ✔
   - Given the array, consider using `string.Join` to combine the list into a single string. ✔

4. Implement the `ISampleData.People` property to return all the items in `People.csv` as `Person` objects ✔

   - Use `ISampleData.CsvRows` as the source of the data. ✔
   - Sort the list by State, City, and Zip. ✔
   - Be sure that `Person.Address` is also populated. ✔
   - Adding null validation to all the `Person` and `Address` properties is **optional**.
   - Consider using `ISampleData.CsvRows` in your test to verify your results. ✔

5. Implement `ISampleDate.FilterByEmailAddress(Predicate<string> filter)` to return a list of names where the email address matches the `filter`. ✔

   - Use `ISampleData.People` for your data source. ✔

6. Implement `ISampleData.GetAggregateListOfStatesGivenPeopleCollection(IEnumerable<IPerson> people)` to return a `string` that contains a **unique**, comma-separated list of states. ✔

   - Use the `people` parameter from `ISampleData.People` property for your data source. ✔
   - At a minimum, use the `System.Linq.Enumerable.`Aggregate` LINQ method to create your result. ✔
   - Don't forget the list should be unique. ✔
   - It is recommended that, at a minimum, you use `ISampleData.GetUniqueSortedListOfStatesGivenCsvRows` to validate your result.

7. Given the implementation of `Node` in Assignment5

- Implement `IEnumerable<T>` to return all the items in the "circle" of items. ✔
- Add an `IEnumberable<T> ChildItems(int maximum)` method to `Node` that returns the remaining items with a maximum number of items returned less than `maximum`.

## Fundamentals

- Place all shared project properties into a `Directory.Build.Props` file.
- Place all shared project items into a `Directory.Build.targets` file.
- Ensure nullable reference types is enabled ✔
- Ensure that you turn on code analysis for all projects(EnableNETAnalyzers) ✔
- Set `LangVersion` and the `TargetFramework` to the latest released versions available (preview versions optional) ✔
- and enabled .NET analyzers for both projects ✔
- For this assignment, consider using `Assert.AreEqual<T>()` (the generic version) ✔
- All of the above should be unit tested ✔
- Choose simplicity over complexity ✔

Looks good!
